### DISCUSSION TITLE
Release process

### DISCUSSION THREAD

**Discussion by igordejanovic** [textX/textX#349] [General]
It is a time to define some lightweight process that will govern textX development and the release process.

Here is a suggestion:
- We are moving to semantic versioning and a standard format to keep changelogs (see [CHANGELOG.md](https://github.com/textX/textX/blob/master/CHANGELOG.md)).
- We develop features on feature branches. Feature branches are named `feature/<feature name>`.
- We fix bugs on bugfix branches named as `bugfix/<issue no>-<name>`
- We have a branch for the upcoming major release -- `next-release` (red in the picture bellow).
- If the feature is backward incompatible (BIC) the PR is made against the `next-release` branch (not against the `master` branch)
- If the feature is backward compatible PR is made against the `master`.
- Thus, `Unreleased` section in the changelog on the `master` branch will never have any BIC change. All BIC changes should go to the changelog on the `next-release` branch.
- We constantly merge `master` branch  to `next-release` branch. Thus, `next-release` branch is the latest and greatest version with **all** finished features applied.
- When the time for major release come we create maintenance branch from the last released version/tag of the form `<major>.<minor>.x` (green in the picture bellow), we merge `next-release` branch to `master` and  tag the new major version release.
  
Here is a drawing that depicts the process:
![image](https://user-images.githubusercontent.com/1673425/58026599-19deb200-7b17-11e9-9869-033f00889a5c.png)

This approach makes releasing minor and bugfix releases possible while developing BICs for the next major release. Users should be safe declaring a dependency towards the major version. Also, following `master` branch should be safe up until a new major release has been made.

**Comment by goto40** [comment]
Very good!

How do we create feature/bugfix branches? Naming conventions?

**Comment by danixeee** [comment]
When we discover and fix a bug that exists in _all_ major releases, how we are going to handle it?
I guess fixed bug can be merged to master and the lts release can be tagged with increased version (e.g. v2.4.2). For non-lts releases we will create a release branch from a tag (e.g. v1.2.3), update changelog, etc and release bugfix from there, am I right?

**Comment by igordejanovic** [comment]
> How do we create feature/bugfix branches? Naming conventions?

We can use `feature/feature_name` and `bugfix/issue...`.

> When we discover and fix a bug that exists in all major releases, how we are going to handle it? I guess fixed bug can be merged to master and the lts release can be tagged with increased version (e.g. v2.4.2). For non-lts releases we will create a release branch from a tag (e.g. v1.2.3), update changelog, etc and release bugfix from there, am I right?

Yes, bugfix branch is merged to master and a new bugfix release is produced by increasing the third digit in the version number, tag is created. I don't think that we should support all major releases but we could support previous major release for some time. If so, we could cherry pick the bugfix to the maintenance branch of the previous version and tag bugfix release there.

For example, lets suppose that the last released version is `2.1.1`, and the last released 1.x version is `1.7.3`. Just before merging 2.x branch to master and making a new major release we create 1.x maintenance branch. 

If we fix a bug, we merge to master and tag it with `2.1.2` to release a new bugfix version. If that bug exists in the 1.x versions we cherry-pick a bugfix to the 1.x maintenance branch and release a new bugfix `1.7.4` version on that branch. 

I have update the drawing above to depict that situation (the maintenance branch is in green).

**Comment by goto40** [comment]
I merged my last PR into master... I am not sure when we start with the new process.

Also I am unsure concerning how to handle "Possible BICs"... According to our plan, possible BICs should also go into the NEXT.VERSION-branch, and not into master?

I hope I did not mess up things with my PRs...

**Comment by Aluriak** [comment]
Just sharing a counter-proposition for the whole branching strategy: [git flow](https://nvie.com/posts/a-successful-git-branching-model).

In short, it is articulated around [two main branches](https://nvie.com/posts/a-successful-git-branching-model/#the-main-branches), dev and master. Master is always releasable, and dev is preparing next version.

Maybe we don't need all the bells and whistles around the tagging and auto-scripting, but i see advantages into using well-documented and well-tested strategy.

**Comment by igordejanovic** [comment]
@goto40 
> I merged my last PR into master... I am not sure when we start with the new process.
>  I hope I did not mess up things with my PRs...

Everything is OK. :) I think we should start with the new release process with the 2.0.0

> Also I am unsure concerning how to handle "Possible BICs"... According to our plan, possible BICs should also go into the NEXT.VERSION-branch, and not into master?

Yes, I think that would be a correct approach. If we suspect that something might cause trouble in upgrading we should schedule it for the next major version release. The idea is to assure the stability of the master branch and all minor version upgrades. User should generally declare version compatibility (dependency) with all versions up to but excluding the next major version which might require manual intervention during upgrade.

> Just sharing a counter-proposition for the whole branching strategy: [git flow](https://nvie.com/posts/a-successful-git-branching-model).
> Maybe we don't need all the bells and whistles around the tagging and auto-scripting, but i see advantages into using well-documented and well-tested strategy.

@Aluriak The model outlined here is in a sense a super-set of gitflow. Gitflow doesn't care about semantic versioning and backward compatibility. The idea here is to try to incorporate that also.

We will have basically two development branches: master and the next major release branch. `master` would incorporate all backward compatible features developed so far and would be a stable branch generally usable by the wide audience. The next major release would have all changes -- BIC features would be merged directly to this branch and we will constantly merge master to this branch to bring all non-BIC features. So, those who wants to test the latest and greatest features could follow the next major release branch. As a matter of fact we could call this branch `develop` until it is released at which point a new release branch is started `2.x`, `3.x`... while the `develop` continues to be the next major release. This will be more similar to gitflow and probably more familiar to contributors used to gitflow. Oh the other hand, it might confuse them. But I think it is generally a good idea to have a stable name for this branch like e.g. `next-major` or something like that. Ideas?

So, to sum things up. With the approach outlined here we could release minor version often, e.g. whenever we merge a new non-BIC feature to master we could release a new minor version. This would enable a fast-track updates without sacrificing the stability. While, with gitflow we would have to hold off with BIC feature merging until we decide to release next major version. At that point we could have a lot of BIC features that could potentially be hard to merge together or would introduce bugs not visible until we start merging those features.

**Comment by igordejanovic** [comment]
I have updated the initial proposal based on the discussion.
